src/t_string.c

```diff
@@ -93,91 +97,181 @@ void setGenericCommand(client *c, int flags, robj *key, robj *val, robj *expire,
         if (getGenericCommand(c) == C_ERR) return;
     }
 
-    genericSetKey(c,c->db,key,val,flags & OBJ_SET_KEEPTTL,1);
+    genericSetKey(c,c->db,key, val,flags & OBJ_KEEPTTL,1);
     server.dirty++;
-    if (expire) setExpire(c,c->db,key,mstime()+milliseconds);
     notifyKeyspaceEvent(NOTIFY_STRING,"set",key,c->db->id);
-    if (expire) notifyKeyspaceEvent(NOTIFY_GENERIC,
-        "expire",key,c->db->id);
+    if (expire) {
+        robj *exp = shared.pxat;
+
+        if ((flags & OBJ_PX) || (flags & OBJ_EX)) {
+            setExpire(c,c->db,key,milliseconds + mstime());
+            exp = shared.px;
+        } else {
+            setExpire(c,c->db,key,milliseconds);
+        }
+        notifyKeyspaceEvent(NOTIFY_GENERIC,
+                            "expire", key, c->db->id);
+
+        /* Propagate as SET Key Value PXAT millisecond-timestamp if there is EXAT/PXAT or
+         * propagate as SET Key Value PX millisecond if there is EX/PX flag.
+         *
+         * Additionally when we propagate the SET with PX (relative millisecond) we translate
+         * it again to SET with PXAT for the AOF.
+         *
+         * Additional care is required while modifying the argument order. AOF relies on the
+         * exp argument being at index 3.
+         * */
+        robj *millisecondObj = createStringObjectFromLongLong(milliseconds);
+        rewriteClientCommandVector(c,5,shared.set,key,val,exp,millisecondObj);
+        decrRefCount(millisecondObj);
+    }
     if (!(flags & OBJ_SET_GET)) {
         addReply(c, ok_reply ? ok_reply : shared.ok);
     }
+
+    /* Propagate without the GET argument */
```

Comment:
```suggestion
    /* Propagate without the GET argument (Isn't needed if we had expire since in that case we completely re-written the command argv) */
```